### PR TITLE
Adapt 1114

### PR DESCRIPTION
**Pull request by bell87**
Added a statement model for open text input.
Open text input uses https://github.com/adlnet/xAPI-Spec/blob/master/xAPI.md#long-fill-in as an interaction type. 

I've merged @ryanlynch7 's branch into this branch to try and avoid merge conflicts later.

### CODE REVIEW

File: js/adapt-xapi.js [hunk 1]
```diff
@@ -93,10 +105,45 @@ define(function(require) {
...
+      var data = {
+        activityId: this.get('activityId'),
+        actor: this.get('actor'),
+        registration: this.get('registration'),
+        model: question.model
+      };
+
+      var statementModel;
+      switch (question.model.get('_component')) {
+        case "mcq":
```
Comment (brian-learningpool) [inline js/adapt-xapi.js]: I'm not sure about this as it means very tight coupling between the question components and the xAPI extension. Is it necessary to have specific statement models when question models expose the same properties anyway?  Could we get around requiring specific implementations per question type if we had the xAPI extension decorate the properties of the question model, and let us know what properties we should be passing?

File: js/models/component-statement.js [hunk 2]
```diff
@@ -34,24 +36,29 @@ define(function(require) {
     },
 
     getVerb: function() {
-      return StatementModel.prototype.getVerb.call(this);
-    },
+      var verb = StatementModel.prototype.getVerb.call(this);
 
-    getObject: function() {
-      return StatementModel.prototype.getObject.call(this);
+      if (
```
Comment (brian-learningpool) [inline js/models/component-statement.js]: I'm not a fan of the brackets like this.

Why not just write:
```javascript
return _.isNull(verb) ? null : verb;
```

File: js/models/component-statement.js [hunk 3]
```diff
@@ -81,22 +88,31 @@ define(function(require) {
...
           var assessmentIri = [this.get('activityId'), 'assessment', article.get('_id')].join('/');
           contextActivities.parent = {
-            id : assessmentIri
+            id: assessmentIri
+          }
+        }
+      } else {
+        // Parent is the course
+        // component -> block -> article -> page -> course
+        var course = this.get('model').getParent().getParent().getParent().getParent();
```
Comment (brian-learningpool) [inline js/models/component-statement.js]: Can't you just use `Adapt.course`?

File: js/models/component-statement.js [hunk 4]
```diff
@@ -34,24 +36,29 @@ define(function(require) {
...
     },
 
-    getActivityDefinitionObject: function() {
-      var object = StatementModel.prototype.getActivityDefinitionObject.call(this);
+    getObject: function() {
+      var object = StatementModel.prototype.getObject.call(this);
 
-      if (_.isEmpty(object)) {
-        object = {};
+      if (
```
Comment (brian-learningpool) [inline js/models/component-statement.js]: As above, keep the `if` on one line.

File: js/models/opentextinput-component-statement.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+      return score;
+    },
+
+    getObject: function() {
+      var object = QuestionComponentStatementModel.prototype.getObject.call(this);
+      var item = this.get('model').attributes;
+      var correctResponses = [];
+
+      object.definition.interactionType = "long-fill-in";
```
Comment (brian-learningpool) [inline js/models/opentextinput-component-statement.js]: Why not use the interactionType already defined on the model?

File: js/models/question-component-statement.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    getStatement: function() {
+      var statement = ComponentStatementModel.prototype.getStatement.call(this);
+
+      var verb = this.getVerb();
+      var object = this.getObject();
+      var context = this.getContext();
+      var result = this.getResult();
+
+      if (
```
Comment (brian-learningpool) [inline js/models/question-component-statement.js]: This `if` statement appears all over the place.  It should probably be abstracted into a common function.

File: js/models/textinput-component-statement.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+      return score;
+    },
+
+    getObject: function() {
+      var object = QuestionComponentStatementModel.prototype.getObject.call(this);
+      var item = this.get('model').get('_items')[0];
+      var correctResponses = [];
+
+      object.definition.interactionType = "fill-in";
```
Comment (brian-learningpool) [inline js/models/textinput-component-statement.js]: Again, why not take the interactionType from the model?

File: js/adapt-xapi.js (hunk 1 above)
Comment (barry-learningpool) [inline js/adapt-xapi.js]: In that case though each specific question component would need to decorate the question model in its own way, so doesn't that mean the tight coupling is unavoidable?

**Comment by moloko** [comment]
seems to me that xAPI question tracking might well be able to take advantage of what was set up for SCORM 1.2 cmi.interaction tracking - AFAIK xAPI was deliberately designed to be extremely similar to SCORM specifically to make this easier (and because the existing data format works very well).

if you look in questionView.js you'll see there are abstract implementations of two functions: `getResponseType` and `getResponse`. All core question components have concrete implementations of these functions that are responsible for returning the right data for their implementation. So, for example, the matching question's `getResponseType` function will return a value of `""matching"`; textInput's implementation will return `"fill-in"`. The `getResponse` function is responsible for returning the answer data in a relatively generic format appropriate to the response type.

There's also an event - `questionView:recordInteraction` that will be broadcast if a question that's not specifically been excluded from interaction tracking gets answered - this allows the author to control for every individual question whether the user's response gets recorded or not. If you're not bothered about this level of control you can just use the `change:_isInteractionComplete` event to be notified of when the question has been answered instead.

In spoor's `AdaptStatefulSession` you'll see all of this brought together in the `onQuestionRecordInteraction` event handler [Lines 120 - 132](https://github.com/adaptlearning/adapt-contrib-spoor/blob/master/js/adapt-stateful-session.js#L120-L132)  that pulls all of the relevant data together in a completely generic manner in just a few lines of code.

Have a look at what this code is doing because I think these functions should be generating data in a format that will let you handle xAPI interaction tracking without having such tightly-coupled code as there is above.

**Comment by brian-learningpool** [comment]
Closing this PR in light of the recent changes to the `develop` branch which takes more of an approach as outlined above without the tight coupling.

**Pull request closed without merging**